// routes/geoLoc.js
var mongo = require('mongodb');
 
var Server = mongo.Server,
    Db = mongo.Db,
    BSON = mongo.BSONPure;
 
var server = new Server('localhost', 27017, {auto_reconnect: true});
db = new Db('geoDb', server);

db.open(function(err, db) {
    if(!err) {
        console.log("Connected to database");
        db.collection('locations', {strict:true}, function(err, collection) {
            if (err) {
                console.log("collection doesn't exist.");
                populateDB();
            }
        });
    }
});
 
exports.findById = function(req, res) {
    var id = req.params.id;
    console.log('Retrieving geoDb: ' + id);
    db.collection('geoDb', function(err, collection) {
        collection.findOne({'_id':new BSON.ObjectID(id)}, function(err, item) {
            res.send(item);
        });
    });
};
 
exports.findAll = function(req, res) {
    console.log(req);
    db.collection('geoDb', function(err, collection) {
        collection.find().toArray(function(err, items) {
            res.send(items);
        });
    });
};
 
exports.addgeoDb = function(req, res) {
    var geoDb = req.body;
    console.log('Adding geoDb: ' + JSON.stringify(geoDb));
    db.collection('geoDb', function(err, collection) {
        collection.insert(geoDb, {safe:true}, function(err, result) {
            if (err) {
                res.send({'error':'An error has occurred'});
            } else {
                console.log('Success: ' + JSON.stringify(result[0]));
                res.send(result[0]);
            }
        });
    });
}
 
exports.updategeoDb = function(req, res) {
    var id = req.params.id;
    var geoDb = req.body;
    console.log('Updating geoDb: ' + id);
    console.log(JSON.stringify(geoDb));
    db.collection('geoDb', function(err, collection) {
        collection.update({'_id':new BSON.ObjectID(id)}, geoDb, {safe:true}, function(err, result) {
            if (err) {
                console.log('Error updating geoDb: ' + err);
                res.send({'error':'An error has occurred'});
            } else {
                console.log('' + result + ' document(s) updated');
                res.send(geoDb);
            }
        });
    });
}
 
exports.deletegeoDb = function(req, res) {
    var id = req.params.id;
    console.log('Deleting geoDb: ' + id);
    db.collection('geoDb', function(err, collection) {
        collection.remove({'_id':new BSON.ObjectID(id)}, {safe:true}, function(err, result) {
            if (err) {
                res.send({'error':'An error has occurred - ' + err});
            } else {
                console.log('' + result + ' document(s) deleted');
                res.send(req.body);
            }
        });
    });
}
 
/*--------------------------------------------------------------------------------------------------------------------*/
// Populate database with sample data -- Only used once: the first time the application is started.
// You'd typically not find this code in a real-life app, since the database would already exist.
var populateDB = function() {
 
    var geoDb = [
    {
        latitude: 51.511214,
        longitude: -0.119824,
        string: 'London'
    },
    {
        latitude: 19.075984,
        longitude: 72.877656,
        string: 'Mumbai'
    }];
 
    db.collection('geoDb', function(err, collection) {
        collection.insert(geoDb, {safe:true}, function(err, result) {});
    });
 
};
